Stremio/stremio-addon-sdk
# P2P addons

Fixes: #69, #48

How to test it:

* run the `supernode`
* run an addon
* run `publish` against the supernode with your addon
* try accessing your addon through the supernode: manifest.json (a file that should already be cached)
* try accessing something that will not be cached, and will be requested on-demand 
* stop the addon, try accessing all the cached content after

TODO documentation of:

* basic spec/documentation of how it works (replacing the previous SIP01 doc, and based on the **new labs#27 doc**)  
* publish: what it is, what it does
* supernode: what it is, what it does
* document that both of them require IPFS
* mnemonic/identity: why it's important
* additional bonus feature that you can use it to convert legacy addons to new ones (being able to replicate legacy addons into a V3-compatible endpoint)
* IPFS, explain immutability and how we solve it
* msg types: Request, Response, Publish

other TODOs:

* **minor attack vector:** someone can keep replaying your last Publish message (if they have access to it), therefore disabling your ability to connect and update your addon (cause of the one-connection-per-identifier limit)
* error "only one connection per addon" needs to be in publish, not in the supernode
* **Ability to deploy a new addon** created by `bootstrap` directly, by having it import `publish` and run it with the `AddonInterface` directly 
* consider renaming `Publish` cause we might add another type of `Publish msg that permanently associates an identifier with an identity
* should we change the manifest ID to reflect the new identifier? probably.
* Can be honeypotted (supernode knows IP of publisher); solutions: hardcode a "trusted" set of supernodes, default behavior to be to deploy to nowsh
* Documentation: when deploying your addon, run it with `NODE_ENV=production` to enable cache headers
* publish to multiple nodes/propagate?
